lib/Core/Memory.cpp

```diff
@@ -798,50 +861,80 @@ ref<Expr> PersistentState::isCacheLinePersisted(unsigned cacheLine,
-                                                              ExecutionState &state, 
-                                                              unsigned cacheLine) const {
-  return getRootCause(solver, state, ConstantExpr::create(cacheLine, Expr::Int32));
+std::unordered_set<std::string> 
+PersistentState::getRootCause(TimingSolver *solver, 
+                              ExecutionState &state,
+                              const UpdateList &ul,
+                              unsigned cacheLine) const {
+  return getRootCause(solver, state, ul, ConstantExpr::create(cacheLine, Expr::Int32));
 }
 
 std::unordered_set<std::string>
 PersistentState::getRootCause(TimingSolver *solver,
                               ExecutionState &state,
+                              const UpdateList &ul,
                               ref<Expr> cacheLine) const {
-  ref<Expr> result = ReadExpr::create(rootCauseLocations,
-                                      ZExtExpr::create(cacheLine, Expr::Int32));
+  ref<Expr> result = ReadExpr::create(ul, ZExtExpr::create(cacheLine, Expr::Int32));
   
   std::unordered_set<std::string> possible;
 
-  for (const auto &p : allRootLocations) {
-    ref<Expr> locId = ConstantExpr::create(p.second, rootCauseLocations.root->range);
-    ref<Expr> isCause = EqExpr::create(result, locId);
-    bool mayBeCause = false;
-    bool success = solver->mayBeTrue(state, isCause, mayBeCause);
-    if (success && mayBeCause) {
-      possible.insert(p.first); // copy
-    } else if (!success) {
-      klee_warning("Did not succeed at finding the root cause!");
-    }
-  }
+  // llvm::errs() << "getRootCause: " << *cacheLine << "\n";
 
-  // If there are no causes, then there must be no updates.
-  if (!possible.size()) {
-    ref<Expr> nullId = ConstantExpr::create(0, rootCauseLocations.root->range);
-    ref<Expr> isNullptr = EqExpr::create(result, getNullptr());
-    bool mustBeNull = false;
-    bool success = solver->mustBeTrue(state, isNullptr, mustBeNull);
+  #if 1
+  std::pair<ref<Expr>, ref<Expr>> range = solver->getRange(state, result);
+  ref<ConstantExpr> lo = dyn_cast<ConstantExpr>(range.first);
+  ref<ConstantExpr> hi = dyn_cast<ConstantExpr>(range.second); 
+  assert(!lo.isNull() && !hi.isNull());
+  if (lo->getZExtValue() == 0 && hi->getZExtValue() == 0) {
+    // llvm::errs() << "all 0\n";
+    return possible;
+  }
 
-    assert(success && "Could not solve!");
-    assert(mustBeNull && "Could not prove null!");
+  for (uint64_t cl = 0; cl < numCacheLines(); ++cl) {
+    ref<Expr> clExpr = ConstantExpr::create(cl, rootCauseWidth);
+    ref<Expr> clVal = ReadExpr::create(ul, 
+                                       ZExtExpr::create(clExpr, Expr::Int32));
+    std::pair<ref<Expr>, ref<Expr>> range = solver->getRange(state, clVal);
+
+    ref<ConstantExpr> lo = dyn_cast<ConstantExpr>(range.first);
+    ref<ConstantExpr> hi = dyn_cast<ConstantExpr>(range.second);  
+    assert(!lo.isNull() && !hi.isNull());
```

Comment:
When are they null? If solving fails? If so, then:

```suggestion
    assert(!lo.isNull() && !hi.isNull() && "FIXME: unhandled solver failure");
```